polymetis/polymetis/python/polymetis/robot_interface.py

```diff
@@ -456,22 +461,104 @@ def move_to_ee_pose(
 
+    """
+    Continuous control methods
+    """
+
+    def start_joint_impedance(self, Kq=None, Kqd=None, **kwargs):
+        """Starts joint position control mode.
+        Runs an non-blocking joint impedance controller.
+        The desired joint positions can be updated using `update_desired_joint_positions`
+        """
+        self._mode = self.ControlMode.JOINT_IMP
+
+        torch_policy = toco.policies.JointImpedanceControl(
+            joint_pos_current=self.get_joint_positions(),
+            Kp=Kq or self.Kq_default,
+            Kd=Kqd or self.Kqd_default,
+            robot_model=self.robot_model,
+            ignore_gravity=self.use_grav_comp,
+        )
+
+        return self.send_torch_policy(torch_policy=torch_policy, blocking=False)
+
+    def start_cartesian_impedance(self, Kx=None, Kxd=None, **kwargs):
+        """Starts Cartesian position control mode.
+        Runs an non-blocking Cartesian impedance controller.
+        The desired EE pose can be updated using `update_desired_ee_pose`
+        """
+        self._mode = self.ControlMode.CARTESIAN_IMP
+
+        torch_policy = toco.policies.CartesianImpedanceControl(
+            joint_pos_current=self.get_joint_positions(),
+            Kp=Kx or self.Kx_default,
+            Kd=Kxd or self.Kxd_default,
+            robot_model=self.robot_model,
+            ignore_gravity=self.use_grav_comp,
+        )
+
+        return self.send_torch_policy(torch_policy=torch_policy, blocking=False)
+
+    def update_desired_joint_positions(self, positions: torch.Tensor):
+        """Update the desired joint positions used by the joint position control mode.
+        Requires starting a joint impedance controller with `start_joint_impedance` beforehand.
+        """
+        if self._mode != self.ControlMode.JOINT_IMP:
+            log.warning(
+                "No joint impedance controller running, use 'start_joint_impedance' to start one."
+            )
+            return
+
+        param_dict = {"joint_pos_desired": positions}
+
+        return self.update_current_policy(param_dict)
+
+    def update_desired_ee_pose(
+        self,
+        position: torch.Tensor = None,
+        orientation: torch.Tensor = None,
+    ):
+        """Update the desired EE pose used by the Cartesian position control mode.
+        Requires starting a Cartesian impedance controller with `start_cartesian_impedance` beforehand.
+        """
+        if self._mode != self.ControlMode.CARTESIAN_IMP:
+            log.warning(
+                "No cartesian impedance controller running, use 'start_cartesian_impedance' to start one."
+            )
+            return
+
+        param_dict = {}
+        if position is not None:
+            param_dict["ee_pos_desired"] = position
+        if orientation is not None:
+            param_dict["ee_quat_desired"] = orientation
+
+        return self.update_current_policy(param_dict)
+
+    def terminate_impedance_controller(self):
```

Comment:
Do we need a separate of impedance controllers specifically?

Reply:
?

Reply:
Separate termination of impedance controllers*

Reply:
I don't think there needs to be. I think even having a single termination method for all non-blocking controllers would've been fine if we don't need to keep track of the mode on client side.